### PR TITLE
Add async setter with concurrency limit

### PR DESCRIPTION
**Pull request by manzt**
Similar to #11, allows user to write to store asynchronously. Uses [`p-queue`](https://github.com/sindresorhus/p-queue) to handle concurrency. 

### Usage
```javascript
const a = new NestedArray(null, 100, "<i4");
const z = await createArray(a.shape, { chunks: 10, dtype: a.dtype });

const setterConfig = { 
    concurrencyLimit: 10, //default
    progressCallback: ({ progress, queueSize }) => console.log(progress/queueSize),
} 
await z.set(null, a, setterConfig);
```

Before merging this, I think we should add tests for both the callback implemented in #11 and here.

### CODE REVIEW

**Comment by manzt** [comment]
Ahh, I'm seeing you added a test in #13. I can add something similar here!

File: src/core/index.ts [hunk 1]
```diff
@@ -459,24 +482,35 @@ export class ZarrArray {
...
+
+      for (const proj of indexer.iter()) {
+        const chunkValue = this.getChunkValue(proj, indexer, value, selectionShape);
+        queue.add(() => this.chunkSetItem(proj.chunkCoords, proj.chunkSelection, chunkValue));
       }
 
-      await this.chunkSetItem(proj.chunkCoords, proj.chunkSelection, chunkValue);
     }
+
+    // gaurantess that all work on queue has finished
```
Comment (gzuidhof) [inline src/core/index.ts]:
```suggestion
    // guarantees that all work on queue has finished
```

File: src/core/index.ts [hunk 2]
```diff
@@ -25,6 +24,14 @@ interface GetOptions {
   }) => void;
 }
 
+interface SetOptions {
```
Comment (gzuidhof) [inline src/core/index.ts]: I think we should actually export these, that allows users of the library to use the typing.

**Comment by manzt** [comment]
Do you mind if I merge this?

**Comment by gzuidhof** [comment]
Go for it when its ready :) Thanks!

**Pull request merged**